controllers/datasource_controller.go

```diff
@@ -166,151 +166,200 @@ func (r *GrafanaDatasourceReconciler) Reconcile(ctx context.Context, req ctrl.Re
-		if err != nil {
-			return ctrl.Result{RequeueAfter: RequeueDelay}, err
-		}
+		// Force requeue for datasource creation
+		return ctrl.Result{Requeue: true}, nil
+	}
 
-		// Status update should trigger the next reconciliation right away, no need to requeue for dashboard creation
-		return ctrl.Result{}, nil
+	datasource, hash, err := r.buildDatasourceModel(ctx, cr)
+	if err != nil {
+		setInvalidSpec(&cr.Status.Conditions, cr.Generation, "InvalidModel", err.Error())
+		meta.RemoveStatusCondition(&cr.Status.Conditions, conditionDatasourceSynchronized)
+		return ctrl.Result{}, fmt.Errorf("could not build datasource model: %w", err)
 	}
 
-	success := true
-	for _, grafana := range instances.Items {
-		// check if this is a cross namespace import
-		if grafana.Namespace != cr.Namespace && !cr.IsAllowCrossNamespaceImport() {
-			continue
-		}
+	removeInvalidSpec(&cr.Status.Conditions)
 
+	pluginErrors := make(map[string]string)
+	applyErrors := make(map[string]string)
+	for _, grafana := range instances {
 		grafana := grafana
-		// an admin url is required to interact with grafana
-		// the instance or route might not yet be ready
-		if grafana.Status.Stage != v1beta1.OperatorStageComplete || grafana.Status.StageStatus != v1beta1.OperatorStageResultSuccess {
-			controllerLog.Info("grafana instance not ready", "grafana", grafana.Name)
-			success = false
-			continue
-		}
 
 		if grafana.IsInternal() {
 			// first reconcile the plugins
 			// append the requested datasources to a configmap from where the
 			// grafana reconciler will pick them upi
 			err = ReconcilePlugins(ctx, r.Client, r.Scheme, &grafana, cr.Spec.Plugins, fmt.Sprintf("%v-datasource", cr.Name))
 			if err != nil {
-				success = false
-				controllerLog.Error(err, "error reconciling plugins", "datasource", cr.Name, "grafana", grafana.Name)
+				pluginErrors[fmt.Sprintf("%s/%s", grafana.Namespace, grafana.Name)] = err.Error()
 			}
 		}
 
 		// then import the datasource into the matching grafana instances
 		err = r.onDatasourceCreated(ctx, &grafana, cr, datasource, hash)
 		if err != nil {
-			success = false
-			cr.Status.LastMessage = err.Error()
-			controllerLog.Error(err, "error reconciling datasource", "datasource", cr.Name, "grafana", grafana.Name)
+			applyErrors[fmt.Sprintf("%s/%s", grafana.Namespace, grafana.Name)] = err.Error()
 		}
 	}
 
-	// if the datasource was successfully synced in all instances, wait for its re-sync period
-	if success {
-		cr.Status.LastMessage = ""
-		cr.Status.Hash = hash
-		if cr.ResyncPeriodHasElapsed() {
-			cr.Status.LastResync = metav1.Time{Time: time.Now()}
-		}
-		cr.Status.UID = cr.CustomUIDOrUID()
-		return ctrl.Result{RequeueAfter: cr.Spec.ResyncPeriod.Duration}, r.Client.Status().Update(ctx, cr)
-	} else {
-		// if there was an issue with the datasource, update the status
-		return ctrl.Result{RequeueAfter: RequeueDelay}, r.Client.Status().Update(ctx, cr)
+	// NOTE New Condition?
+	// Specific to datasources
```

Comment:
A dedicated status would be nice but I'd implement this in a separate PR

Reply:
This would make a lot of sense to do once Dashboard and the Grafana reconciler have also been updated with Conditions.